ethcore/light/src/on_demand/mod.rs

```diff
@@ -435,82 +471,65 @@ impl OnDemand {
 	// dispatch pending requests, and discard those for which the corresponding
 	// receiver has been dropped.
 	fn dispatch_pending(&self, ctx: &BasicContext) {
-		if self.pending.read().is_empty() { return }
-		let mut pending = self.pending.write();
+		if self.pending.read().is_empty() {
+			return
+		}
 
-		debug!(target: "on_demand", "Attempting to dispatch {} pending requests", pending.len());
+		let mut pending = self.pending.write();
 
 		// iterate over all pending requests, and check them for hang-up.
 		// then, try and find a peer who can serve it.
 		let peers = self.peers.read();
-		*pending = ::std::mem::replace(&mut *pending, Vec::new()).into_iter()
+
+		*pending = ::std::mem::replace(&mut *pending, Vec::new())
+			.into_iter()
 			.filter(|pending| !pending.sender.is_canceled())
 			.filter_map(|mut pending| {
-				// the peer we dispatch to is chosen randomly
+
 				let num_peers = peers.len();
-				let history_len = pending.query_id_history.len();
-				let offset = if history_len == 0 {
-					pending.remaining_query_count = self.base_retry_count;
-					let rand = rand::random::<usize>();
-					pending.base_query_index = rand;
-					rand
-				} else {
-					pending.base_query_index + history_len
-				} % cmp::max(num_peers, 1);
-				let init_remaining_query_count = pending.remaining_query_count; // to fail in case of big reduction of nb of peers
-				for (peer_id, peer) in peers.iter().chain(peers.iter())
-					.skip(offset).take(num_peers) {
-					// TODO: see which requests can be answered by the cache?
-					if pending.remaining_query_count == 0 {
-						break
+				// The first peer to dispatch the request is chosen at random
+				let rand = rand::random::<usize>() % cmp::max(num_peers, 1);
+
+				for (peer_id, peer) in peers
+					.iter()
+					.cycle()
```

Comment:
Is the switch to 'cycle' needed here, I mean it feels to me like previous 'chain' call does ensure we did not infiniloop (I am not saying we can (I did not check) but just that it makes it easier to be sure of the fact). I do not think there is a need to loop more than once other all peers to find next dispatch.

Reply:
`.take(num_peers)` ensures we don't have infiniloop
`.iter().cycle().skip(rand).take(num_peers)` essentially rotates peers by `rand`